README.md
# puppetdb-ruby

[![Build Status](https://img.shields.io/travis/voxpupuli/puppetdb-ruby.svg)](https://travis-ci.org/voxpupuli/puppetdb-ruby)
[![Gem Version](https://img.shields.io/gem/v/puppetdb-ruby.svg)](https://rubygems.org/gems/puppetdb-ruby)
[![Gem Downloads](https://img.shields.io/gem/dt/puppetdb-ruby.svg)](https://rubygems.org/gems/puppetdb-ruby)
[![By VoxPupuli](https://img.shields.io/badge/voxpupuli-%F0%9F%90%B1-orange.svg)](https://voxpupuli.org/)

a simple gem for interacting with the
[PuppetDB](https://github.com/puppetlabs/puppetdb) API.

This library was migrated from [puppetlabs](https://github.com/puppetlabs)
ownership to VoxPupuli on 19 October 2016.

## Installation

Installing from Ruby CLI:
```
gem install puppetdb-ruby
```

Include in Gemfile:
``` ruby
gem 'puppetdb-ruby'
```

## Usage

Require the puppetdb gem in your ruby code.

```ruby
require 'puppetdb'

# Defaults to latest API version.
```

#### Create a new connection:

Non-SSL:
``` ruby
client = PuppetDB::Client.new({:server => 'http://localhost:8080'})
```

SSL with cert-based authentication:
``` ruby
client = PuppetDB::Client.new({
    :server => 'https://localhost:8081',
    :pem    => {
        'key'     => "keyfile",
        'cert'    => "certfile",
        'ca_file' => "cafile"
    }})
```

SSL with PE RBAC token based authentication:
``` ruby
client = PuppetDB::Client.new({
    :server => "https://localhost:8081",
    :token  => "my_pe_rbac_token",
    :cacert => "/path/to/cacert.pem",
    })
```

Configure connections to multiple PuppetDB's via `server_urls`
``` ruby
client = PuppetDB::Client.new({
    :server_urls => "https://localhost:8081,https://localhost:8083",
    :token  => "my_pe_rbac_token",
    :cacert => "/path/to/cacert.pem",
    })
```

SSL with PE RBAC token based authentication, using all settings from PE Client Tools configurations:
``` ruby
client = PuppetDB::Client.new()
```

Note: When using cert-based authentication you must specify the full pem structure. When using token based authentication
you must NOT provide the pem structure and instead pass ':token' and ':cacert' (or allow them to be read from the
PE Client Tools configuration).

#### Query API usage

The Query Feature allows the user to request data from PuppetDB using the Query endpoints. It defaults to the latest version of the Query Endpoint.

Currently, `puppetdb-ruby` only supports the [AST Query Language](https://docs.puppet.com/puppetdb/5.0/api/query/v4/ast.html).

Support for the [PQL Query Language](https://docs.puppet.com/puppetdb/5.0/api/query/tutorial-pql.html) is planned for a future release.

Example:
``` ruby
response = client.request(
  'nodes',
  [:and,
    [:'=', ['fact', 'kernel'], 'Linux'],
    [:>, ['fact', 'uptime_days'], 30]
  ],
  {:limit => 10}
)

nodes = response.data

# queries are composable

uptime = PuppetDB::Query[:>, [:fact, 'uptime_days'], 30]
redhat = PuppetDB::Query[:'=', [:fact, 'osfamily'], 'RedHat']
debian = PuppetDB::Query[:'=', [:fact, 'osfamily'], 'Debian']

client.request uptime.and(debian)
client.request uptime.and(redhat)
client.request uptime.and(debian.or(redhat))
```

If you have configured multiple PuppetDB's via [`server_urls`](https://puppet.com/docs/puppetdb/latest/pdb_client_tools.html#step-3-install-and-configure-the-puppetdb-cli)
then you can query in `:failover` mode. This will query each PuppetDB in `server_urls`
in order until it gets a successful response. It will fail with an `APIError` only if all queries fail.

``` ruby
response = client.request(
  'nodes',
  [:"=", "certname", "foo"],
  {
    :limit => 10
    :query_mode => :failover
  }
)
```

See the [PuppetDB API Docs](https://docs.puppet.com/puppetdb/5.0/api/index.html) for more.


#### PQL Queries usage

PQL queries are supported by using the empty endpoint.

Example:
``` ruby
response = client.request(
  '',
  'resources[title] { nodes { deactivated is null } }',
  {:limit => 10}
)

resources = response.data
```

See the [PuppetDB API Docs](https://docs.puppet.com/puppetdb/5.0/api/query/v4/pql.html) for more on PQL queries.


#### Command API Usage

The Command Feature allows the user to execute REST Commands against the PuppetDB Command API Endpoints. It defaults to the latest version of the Command Endpoint.

The command method takes three arguments:

* `command`: a string identifying the command
* `payload`: a valid JSON object of any sort. It’s up to an individual handler function to determine how to interpret that object.
* `version`: a JSON integer describing what version of the given command you’re attempting to invoke. The version of the command also indicates the version of the wire format to use for the command.

Example:
``` ruby
client.command(
  'deactivate node',
  {'certname' => 'test1', 'producer_timestamp' => '2015-01-01'},
  3
)
```

See the PuppetDB [Commands Endpoint Docs](https://docs.puppet.com/puppetdb/5.0/api/command/v1/commands.html) for more information.

#### Query the status endpoint(s)

You can get the status of all configured PuppetDB's by querying the `/status/v1/services` endpoints.
``` ruby
client.status

# The result will be of the form (one entry per server)
# {
#   "http://localhost:8080": {
#     "puppetdb-status": {
#       "service_version": "6.3.1-SNAPSHOT",
#       "service_status_version": 1,
#       "detail_level": "info",
#       "state": "running",
#       "status": {
#         "maintenance_mode?": false,
#         "queue_depth": 0,
#         "read_db_up?": true,
#         "write_db_up?": true
#       },
#       "active_alerts": [
#       ]
#     },
#     "status-service": {
#       "service_version": "1.1.0",
#       "service_status_version": 1,
#       "detail_level": "info",
#       "state": "running",
#       "status": {
#       },
#       "active_alerts": [
#       ]
#     }
#   }
# }
```

#### Export an archive of PuppetDB

You can [`export`](https://puppet.com/docs/puppetdb/5.1/anonymization.html#using-the-export-command) an archive
of your PuppetDB to a file. Optionally you can override the `anonymization_profile` (default: none).

``` ruby
client.export('path/for/new_puppetdb_export.tar.gz', anonymization_profile: :high)
```

#### Import an archive into PuppetDB

Once you have a PuppetDB export, it can be loaded into PuppetDB with [`import`](https://puppet.com/docs/puppetdb/5.1/anonymization.html#using-the-import-command).

``` ruby
client.import('path/to/existing_puppetdb_export.tar.gz')
```

## Tests

```
bundle install
bundle exec rspec
```

## Issues & Contributions

File issues or feature requests using [GitHub
issues](https://github.com/voxpupuli/puppetdb-ruby/issues).

If you are interested in contributing to this project, please see the
[Contribution Guidelines](CONTRIBUTING.md)

## Authors

This module was donated to VoxPupuli by Puppet Inc on 10-19-2016.

Nathaniel Smith <nathaniel@puppetlabs.com>
Lindsey Smith <lindsey@puppetlabs.com>
Ruth Linehan <ruth@puppetlabs.com>

## License

See LICENSE.


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.
Each new release typically also includes the latest modulesync defaults.
These should not impact the functionality of the module.

## [1.2.0](https://github.com/voxpupuli/puppetdb-ruby/tree/1.2.0) (2019-08-06)
[Full Changelog](https://github.com/voxpupuli/puppetdb-ruby/compare/1.1.1...1.2.0)

**Implemented enhancements:**

- Add PE RBAC token support [\#34](https://github.com/voxpupuli/puppetdb-ruby/pull/34) ([seanmil](https://github.com/seanmil))

**Closed issues:**

- cannot load such file -- puppetdb [\#33](https://github.com/voxpupuli/puppetdb-ruby/issues/33)
- Attempting to use puppetdb-ruby in a custom function [\#11](https://github.com/voxpupuli/puppetdb-ruby/issues/11)
- No way to retrieve all facts/nodes [\#8](https://github.com/voxpupuli/puppetdb-ruby/issues/8)

## [1.1.1](https://github.com/voxpupuli/puppetdb-ruby/tree/1.1.1) (2017-08-17)
[Full Changelog](https://github.com/voxpupuli/puppetdb-ruby/compare/1.1.0...1.1.1)

**Fixed bugs:**

- Rubygem is broken for 1.1.0 [\#31](https://github.com/voxpupuli/puppetdb-ruby/issues/31)

**Merged pull requests:**

- Add badges [\#30](https://github.com/voxpupuli/puppetdb-ruby/pull/30) ([raphink](https://github.com/raphink))

## [1.1.0](https://github.com/voxpupuli/puppetdb-ruby/tree/1.1.0) (2017-08-17)
[Full Changelog](https://github.com/voxpupuli/puppetdb-ruby/compare/1.0.0...1.1.0)

**Closed issues:**

- Improve documentation  [\#22](https://github.com/voxpupuli/puppetdb-ruby/issues/22)
- support for PQL in PuppetDB::Query [\#16](https://github.com/voxpupuli/puppetdb-ruby/issues/16)
- Any plans to revive this project? [\#12](https://github.com/voxpupuli/puppetdb-ruby/issues/12)

**Merged pull requests:**

- Add PQL support [\#28](https://github.com/voxpupuli/puppetdb-ruby/pull/28) ([raphink](https://github.com/raphink))

## [1.0.0](https://github.com/voxpupuli/puppetdb-ruby/tree/1.0.0) (2017-08-02)
**Implemented enhancements:**

- Preparatory cleanup for refactor [\#21](https://github.com/voxpupuli/puppetdb-ruby/pull/21) ([dhollinger](https://github.com/dhollinger))
- Allow usage of unauthenticated PuppetDB over SSL [\#14](https://github.com/voxpupuli/puppetdb-ruby/pull/14) ([cassianoleal](https://github.com/cassianoleal))

**Closed issues:**

- Refactor to support v4 endpoints [\#20](https://github.com/voxpupuli/puppetdb-ruby/issues/20)
- API V4 Example [\#13](https://github.com/voxpupuli/puppetdb-ruby/issues/13)
- 'summarize-by' parameter not supported [\#9](https://github.com/voxpupuli/puppetdb-ruby/issues/9)
- Thanks! [\#6](https://github.com/voxpupuli/puppetdb-ruby/issues/6)
- Only GET requests supported [\#4](https://github.com/voxpupuli/puppetdb-ruby/issues/4)

**Merged pull requests:**

- rename gemspec to fix deploy [\#27](https://github.com/voxpupuli/puppetdb-ruby/pull/27) ([dhollinger](https://github.com/dhollinger))
- Remove branches restriction in travis.yml [\#26](https://github.com/voxpupuli/puppetdb-ruby/pull/26) ([dhollinger](https://github.com/dhollinger))
- Add rubygems deploy key [\#25](https://github.com/voxpupuli/puppetdb-ruby/pull/25) ([dhollinger](https://github.com/dhollinger))
- Release 1.0.0 [\#24](https://github.com/voxpupuli/puppetdb-ruby/pull/24) ([dhollinger](https://github.com/dhollinger))
- Update Docs for Release 1.0.0 [\#23](https://github.com/voxpupuli/puppetdb-ruby/pull/23) ([dhollinger](https://github.com/dhollinger))
- Add command API support [\#19](https://github.com/voxpupuli/puppetdb-ruby/pull/19) ([bekbulatov](https://github.com/bekbulatov))
- \(maint\) Transfer ownership to Vox Pupuli [\#18](https://github.com/voxpupuli/puppetdb-ruby/pull/18) ([rlinehan](https://github.com/rlinehan))
- "Voxpupulify" [\#17](https://github.com/voxpupuli/puppetdb-ruby/pull/17) ([nibalizer](https://github.com/nibalizer))
- Add Contributing Guidelines [\#7](https://github.com/voxpupuli/puppetdb-ruby/pull/7) ([rlinehan](https://github.com/rlinehan))
- Feature/fix readme [\#3](https://github.com/voxpupuli/puppetdb-ruby/pull/3) ([robinbowes](https://github.com/robinbowes))
- Update README with correct initialization of client [\#1](https://github.com/voxpupuli/puppetdb-ruby/pull/1) ([ccaum](https://github.com/ccaum))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


spec/spec_helper.rb
RSpec.configure do |conf|
  conf.mock_framework = :mocha
end


spec/unit/query_spec.rb
require File.join(File.dirname(__FILE__), '..', 'spec_helper')
require File.join(File.dirname(__FILE__), '..', '..', 'lib', 'puppetdb')

def make_mock_response
  m = mock
  m.stubs(:code).returns(200)
  m.expects(:parsed_response).returns(['foo'])
  m
end

def make_mock_query
  m = mock
  m.expects(:build)
  m.expects(:summarize_by).returns(m)
  m
end

def expect_include_total(mock_query)
  mock_query.expects(:include_total).with(true)
  mock_query
end

describe 'raise_if_error' do
  settings = { 'server' => 'http://localhost:8080' }

  it 'works with 4xx' do
    response = mock
    response.stubs(:code).returns(400)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should raise_error(RuntimeError)
  end

  it 'works with 5xx' do
    response = mock
    response.stubs(:code).returns(500)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should raise_error(RuntimeError)
  end

  it 'ignores 2xx' do
    response = mock
    response.stubs(:code).returns(200)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should_not raise_error
  end

  it 'ignores 3xx' do
    response = mock
    response.stubs(:code).returns(300)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should_not raise_error
  end
end

describe 'SSL support' do
  describe 'when http:// is specified' do
    it 'does not use ssl' do
      settings = {
        'server' => 'http://localhost:8080'
      }

      r = PuppetDB::Client.new(settings)
      expect(r.use_ssl).to eq(false)
    end
  end

  describe 'when https:// is specified' do
    it 'uses ssl' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem' => {
          'cert'    => 'foo',
          'key'     => 'bar',
          'ca_file' => 'baz'
        }
      }

      r = PuppetDB::Client.new(settings)
      expect(r.use_ssl).to eq(true)
    end

    it 'tolerates lack of pem' do
      settings = {
        server: 'https://localhost:8081'
      }

      -> { PuppetDB::Client.new(settings) }.should_not raise_error
    end

    it 'does not tolerate lack of key' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem'    => {
          'cert' => 'foo',
          'ca_file' => 'bar'
        }
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end

    it 'does not tolerate lack of cert' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem'    => {
          'key' => 'foo',
          'ca_file' => 'bar'
        }
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end

    it 'does not tolerate lack of ca_file' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem'    => {
          'key' => 'foo',
          'cert' => 'bar'
        }
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end
  end

  describe 'when a protocol is missing from config file' do
    it 'raises an exception' do
      settings = {
        'server' => 'localhost:8080'
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end
  end
end

describe 'request' do
  settings = { server: 'http://localhost' }

  it 'works with array instead of Query' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:headers).returns('X-Records' => 0)
    mock_response.expects(:parsed_response).returns([])

    PuppetDB::Client.expects(:get).returns(mock_response).at_least_once.with do |_path, opts|
      opts[:body] == { 'query' => '[1,2,3]' }
    end
    client.request('/foo', [1, 2, 3])
  end

  it 'processes options correctly' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:headers).returns('X-Records' => 0)
    mock_response.expects(:parsed_response).returns([])

    PuppetDB::Client.expects(:get).returns(mock_response).at_least_once.with do |_path, opts|
      opts == {
        body: {
          'query'         => '[1,2,3]',
          'limit'         => 10,
          'counts-filter' => '[4,5,6]',
          'foo-bar'       => 'foo'
        }
      }
    end

    client.request('/foo', PuppetDB::Query[1, 2, 3], limit: 10,
                                                     counts_filter: [4, 5, 6],
                                                     foo_bar: 'foo')
  end
end


spec/unit/client_spec.rb
require File.join(File.dirname(__FILE__), '..', 'spec_helper')
require File.join(File.dirname(__FILE__), '..', '..', 'lib', 'puppetdb')
require 'tempfile'

def make_mock_response
  m = mock
  m.stubs(:code).returns(200)
  m.expects(:parsed_response).returns(['foo'])
  m
end

def make_mock_query
  m = mock
  m.expects(:build)
  m.expects(:summarize_by).returns(m)
  m
end

def expect_include_total(mock_query)
  mock_query.expects(:include_total).with(true)
  mock_query
end

describe 'raise_if_error' do
  settings = { 'server' => 'http://localhost:8080' }

  it 'works with 4xx' do
    response = mock
    response.stubs(:code).returns(400)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should raise_error(PuppetDB::APIError)
  end

  it 'works with 5xx' do
    response = mock
    response.stubs(:code).returns(500)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should raise_error(PuppetDB::APIError)
  end

  it 'raises UnauthorizedError with 401' do
    response = mock
    response.stubs(:code).returns(401)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should raise_error(PuppetDB::UnauthorizedError)
  end

  it 'raises ForbiddenError with 403' do
    response = mock
    response.stubs(:code).returns(403)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should raise_error(PuppetDB::ForbiddenError)
  end

  it 'ignores 2xx' do
    response = mock
    response.stubs(:code).returns(200)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should_not raise_error
  end

  it 'ignores 3xx' do
    response = mock
    response.stubs(:code).returns(300)

    -> { PuppetDB::Client.new(settings).raise_if_error(response) }.should_not raise_error
  end
end

describe 'SSL support' do
  describe 'when http:// is specified' do
    it 'does not use ssl' do
      settings = {
        'server' => 'http://localhost:8080'
      }

      r = PuppetDB::Client.new(settings)
      expect(r.use_ssl).to eq(false)
    end
  end

  describe 'when https:// is specified' do
    it 'uses ssl' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem' => {
          'cert'    => 'foo',
          'key'     => 'bar',
          'ca_file' => 'baz'
        }
      }

      r = PuppetDB::Client.new(settings)
      expect(r.use_ssl).to eq(true)
    end

    it 'tolerates lack of pem' do
      settings = {
        server: 'https://localhost:8081'
      }

      -> { PuppetDB::Client.new(settings) }.should_not raise_error
    end

    it 'does not tolerate lack of key' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem'    => {
          'cert' => 'foo',
          'ca_file' => 'bar'
        }
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end

    it 'does not tolerate lack of cert' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem'    => {
          'key' => 'foo',
          'ca_file' => 'bar'
        }
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end

    it 'does not tolerate lack of ca_file' do
      settings = {
        'server' => 'https://localhost:8081',
        'pem'    => {
          'key' => 'foo',
          'cert' => 'bar'
        }
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end

    context 'when using token auth' do
      settings = {
        'server' => 'https://localhost:8081'
      }

      before do
        Dir.stubs(:home).returns('/user/root')
        File.stubs(:readable?).with('/user/root/.puppetlabs/token').returns(true)
        File.stubs(:read).with('/user/root/.puppetlabs/token').returns('mytoken')
      end

      it 'does not raise an error when no token or pem is provided' do
        -> { PuppetDB::Client.new(settings) }.should_not raise_error
      end

      it 'configures the header with the token' do
        r = PuppetDB::Client.new(settings)
        expect(r.class.headers).to include('X-Authentication' => 'mytoken')
      end

      it 'will set an empty pem' do
        r = PuppetDB::Client.new(settings)
        expect(r.class.default_options).to include(pem: {})
      end

      it 'uses the default cacert path' do
        r = PuppetDB::Client.new(settings)
        expect(r.class.default_options).to include(cacert: '/etc/puppetlabs/puppet/ssl/certs/ca.pem')
      end

      it 'will use a provided cacert path' do
        r = PuppetDB::Client.new(settings.merge('cacert' => '/my/ca/path'))
        expect(r.class.default_options).to include(cacert: '/my/ca/path')
      end
    end
  end

  describe 'when a protocol is missing from config file' do
    it 'raises an exception' do
      settings = {
        'server' => 'localhost:8080'
      }

      -> { PuppetDB::Client.new(settings) }.should raise_error(RuntimeError)
    end
  end
end

describe 'request' do
  settings = { server: 'http://localhost' }

  it 'works with array instead of Query' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:headers).returns('X-Records' => 0)
    mock_response.expects(:parsed_response).returns([])

    PuppetDB::Client.expects(:get).returns(mock_response).at_least_once.with do |_path, opts|
      opts[:body] == { 'query' => '[1,2,3]' }
    end
    client.request('/foo', [1, 2, 3])
  end

  it 'processes options correctly' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:headers).returns('X-Records' => 0)
    mock_response.expects(:parsed_response).returns([])

    PuppetDB::Client.expects(:get).returns(mock_response).at_least_once.with do |_path, opts|
      opts == {
        body: {
          'query'         => '[1,2,3]',
          'limit'         => 10,
          'counts-filter' => '[4,5,6]',
          'foo-bar'       => 'foo'
        }
      }
    end

    client.request('/foo', PuppetDB::Query[1, 2, 3], limit: 10,
                                                     counts_filter: [4, 5, 6],
                                                     foo_bar: 'foo')
  end

  it 'supports pql' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:headers).returns('X-Records' => 0)
    mock_response.expects(:parsed_response).returns([])

    PuppetDB::Client.expects(:get).returns(mock_response).at_least_once.with do |_path, opts|
      opts == {
        body: {
          'query'         => 'resources{}'
        }
      }
    end

    client.request('', 'resources{}')
  end

  describe 'failover mode' do
    it 'returns successful result' do
      client = PuppetDB::Client.new(server_urls: 'http://localhost:8080,http://localhost:8081')

      mock_response = mock
      mock_response.expects(:code).at_least_once.returns(200)
      mock_response.expects(:headers).returns('X-Records' => 0)
      mock_response.expects(:parsed_response).returns([])

      PuppetDB::Client.expects(:get).returns(mock_response).once.with do |_path, opts|
        opts == {
          body: {
            'query'         => 'resources{}'
          }
        }
      end

      client.request('', 'resources{}', query_mode: :failover)
    end

    it 'throws APIError if all queries fail' do
      client = PuppetDB::Client.new(server_urls: 'http://localhost:8080,http://localhost:8081')

      mock_response = mock
      mock_response.expects(:code).at_least_once.returns(400)

      PuppetDB::Client.expects(:get).returns(mock_response).twice.with do |_path, opts|
        opts == {
          body: {
            'query'         => 'resources{}'
          }
        }
      end

      -> { client.request('', 'resources{}', query_mode: :failover) }.should raise_error(PuppetDB::APIError)
    end
  end
end

describe 'command' do
  settings = { server: 'http://localhost' }
  command = 'deactivate node'
  payload = { 'certname' => 'test1', 'producer_timestamp' => '2015-01-01' }
  payload_version = 3

  it 'processes options correctly' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:parsed_response).returns([])

    PuppetDB::Client.expects(:post).returns(mock_response).at_least_once.with do |_path, opts|
      expect(opts).to eq(query: {
                           'command' => command,
                           'version' => payload_version,
                           'certname' => 'test1'
                         },
                         body: payload.to_json,
                         headers: {
                           'Accept' => 'application/json',
                           'Content-Type' => 'application/json'
                         })
    end
    client.command(command, payload, payload_version)
  end
end

describe 'status' do
  settings = { server: 'http://localhost' }
  two_servers = { server_urls: 'http://localhost:8080,http://localhost:8081' }

  it 'works with one server' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:parsed_response).returns(status: 'running')

    PuppetDB::Client.expects(:get).returns(mock_response).once.with do |path, _opts|
      path == '/status/v1/services'
    end
    expect(client.status).to eq('http://localhost' => { status: 'running' })
  end

  it 'replaces error response with generic message' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(400)

    PuppetDB::Client.expects(:get).returns(mock_response).once.with do |path, _opts|
      path == '/status/v1/services'
    end
    expect(client.status).to eq('http://localhost' => { error: 'Unable to build JSON object from server: http://localhost' })
  end

  it 'queries and aggregates all server statuses' do
    client = PuppetDB::Client.new(two_servers)

    mock_response = mock
    mock_response.expects(:code).at_least_once.returns(200)
    mock_response.expects(:parsed_response).twice.returns(status: 'running')

    PuppetDB::Client.expects(:get).returns(mock_response).twice.with do |path, _opts|
      path == '/status/v1/services'
    end
    expect(client.status).to eq(
      'http://localhost:8080' => { status: 'running' },
      'http://localhost:8081' => { status: 'running' }
    )
  end
end

describe 'import' do
  settings = { server: 'http://localhost' }

  it 'send a multipart POST of the tar' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock

    mock_file = mock
    File.expects(:open).with('exported_pdb_data.tar.gz').returns(mock_file)

    PuppetDB::Client.expects(:post).returns(mock_response).once.with do |path, opts|
      path == '/pdb/admin/v1/archive' &&
        opts[:body][:archive] == mock_file
    end
    client.import 'exported_pdb_data.tar.gz'
  end
end

describe 'export' do
  settings = { server: 'http://localhost' }

  it 'streams body to a file' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    file = Tempfile.new 'export_pdb_data.tar.gz'

    PuppetDB::Client.expects(:get).returns(mock_response).once.with do |path, opts|
      path == '/pdb/admin/v1/archive' &&
        opts[:anonymization_profile] = :none &&
                                       opts[:stream_body] == true
    end
    client.export file.path
  end

  it 'allows customizing the anonymization profile' do
    client = PuppetDB::Client.new(settings)

    mock_response = mock
    file = Tempfile.new 'export_pdb_data.tar.gz'

    PuppetDB::Client.expects(:get).returns(mock_response).once.with do |path, opts|
      path == '/pdb/admin/v1/archive' &&
        opts[:anonymization_profile] = :full &&
                                       opts[:stream_body] == true
    end
    client.export(file.path, anonymization_profile: :full)
  end
end


lib/puppetdb/query.rb
require 'json'

module PuppetDB
  class Query
    attr_reader :sexpr

    def initialize(query = [])
      @sexpr = query
    end

    def self.[](*args)
      Query.new(args)
    end

    def self.maybe_promote(query)
      return Query.new(query) unless query.class == Query
      query
    end

    def empty?
      @sexpr.empty?
    end

    def compose(query)
      query = self.class.maybe_promote(query)

      # If an operand is the empty query ([]), compose returns a copy
      # of the non-empty operand. If both operands are empty, the
      # empty query is returned. If both operands are non-empty, the
      # compose continues.
      if query.empty? && !empty?
        Query.new(@sexpr)
      elsif empty? && !query.empty?
        Query.new(query.sexpr)
      elsif empty? && query.empty?
        Query.new([])
      else
        yield query
      end
    end

    def and(query)
      compose(query) { |q| Query.new([:and, @sexpr, q.sexpr]) }
    end

    def or(query)
      compose(query) { |q| Query.new([:or, @sexpr, q.sexpr]) }
    end

    def push(query)
      compose(query) { |q| Query.new(@sexpr.dup.push(q.sexpr)) }
    end

    def build
      JSON.dump(@sexpr)
    end
  end
end


lib/puppetdb/client.rb
require 'httparty'
require 'logger'

require 'puppetdb/error'

module PuppetDB
  class FixSSLConnectionAdapter < HTTParty::ConnectionAdapter
    def attach_ssl_certificates(http, options)
      if options[:pem].empty?
        http.ca_file = options[:cacert]
      else
        http.cert    = OpenSSL::X509::Certificate.new(File.read(options[:pem]['cert']))
        http.key     = OpenSSL::PKey::RSA.new(File.read(options[:pem]['key']))
        http.ca_file = options[:pem]['ca_file']
      end
      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
    end
  end

  class Client
    include HTTParty
    attr_reader :use_ssl
    attr_writer :logger

    def hash_includes?(hash, *sought_keys)
      sought_keys.each { |x| return false unless hash.include?(x) }
      true
    end

    def debug(msg)
      @logger.debug(msg) if @logger
    end

    def initialize(settings = {}, query_api_version = 4, command_api_version = 1, admin_api_version = 1)
      config = Config.new(settings, load_files: true)
      @query_api_version = query_api_version
      @command_api_version = command_api_version
      @admin_api_version = admin_api_version

      @servers = config.server_urls
      pem    = config['pem'] || {}
      token  = config.token

      @servers.each do |server|
        scheme = URI.parse(server).scheme
        @use_ssl ||= scheme == 'https'

        unless %w[http https].include? scheme
          error_msg = "Configuration error: server_url '#{server}' must specify a protocol of either http or https"
          raise error_msg
        end
      end

      return unless @use_ssl
      unless pem.empty? || hash_includes?(pem, 'key', 'cert', 'ca_file')
        error_msg = 'Configuration error: https:// specified with pem, but pem is incomplete. It requires cert, key, and ca_file.'
        raise error_msg
      end

      self.class.default_options = { pem: pem, cacert: config['cacert'] }
      self.class.headers('X-Authentication' => token) if token
      self.class.connection_adapter(FixSSLConnectionAdapter)
    end

    def raise_if_error(response)
      raise UnauthorizedError, response if response.code == 401
      raise ForbiddenError, response if response.code == 403
      raise APIError, response if response.code.to_s =~ %r{^[4|5]}
    end

    def request(endpoint, query, opts = {})
      path = "/pdb/query/v#{@query_api_version}"
      if endpoint == ''
        # PQL
        json_query = query
      else
        path += "/#{endpoint}"
        query = PuppetDB::Query.maybe_promote(query)
        json_query = query.build
      end

      query_mode = opts.delete(:query_mode) || :first
      filtered_opts = { 'query' => json_query }
      opts.each do |k, v|
        if k == :counts_filter
          filtered_opts['counts-filter'] = JSON.dump(v)
        else
          filtered_opts[k.to_s.sub('_', '-')] = v
        end
      end

      debug("#{path} #{json_query} #{opts}")

      if query_mode == :first
        self.class.base_uri(@servers.first)
        ret = self.class.get(path, body: filtered_opts)
        raise_if_error(ret)

        total = ret.headers['X-Records']
        total = ret.parsed_response.length if total.nil?

        Response.new(ret.parsed_response, total)
      elsif query_mode == :failover

        ret=nil
        @servers.each do |server|
          self.class.base_uri(server)
          ret = self.class.get(path, body: filtered_opts)
          if ret.code < 400
            total = ret.headers['X-Records']
            total = ret.parsed_response.length if total.nil?

            return Response.new(ret.parsed_response, total)
          else
            debug("query on '#{server}' failed with #{ret.code}")
          end
        end
        raise APIError, ret
      else
        raise ArgumentError, "Query mode '#{query_mode}' is not supported (try :first or :failover)."
      end
    end

    def command(command, payload, version)
      path = "/pdb/cmd/v#{@command_api_version}"

      query = {
        'command' => command,
        'version' => version,
        'certname' => payload['certname']
      }

      debug("#{path} #{query} #{payload}")

      self.class.base_uri(@servers.first)
      ret = self.class.post(
        path,
        query: query,
        body: payload.to_json,
        headers: {
          'Accept'       => 'application/json',
          'Content-Type' => 'application/json'
        }
      )
      raise_if_error(ret)

      Response.new(ret.parsed_response)
    end

    def export(filename, opts = {})
      self.class.base_uri(@servers.first)
      path = "/pdb/admin/v#{@admin_api_version}/archive"

      # Allow opts to override anonymization_profile, but enforce
      # stream_body to avoid using memory
      params = { anonymization_profile: 'none' }.
               merge(opts).
               merge(stream_body: true)

      File.open(filename, 'w') do |file|
        self.class.get(path, params) do |fragment|
          if [301, 302].include?(fragment.code)
            debug 'Skip streaming write for redirect'
          elsif fragment.code == 200
            file.write(fragment)
          else
            raise StandardError, "Non-success status code while streaming #{fragment.code}"
          end
        end
      end
    end

    def import(filename)
      self.class.base_uri(@servers.first)
      path = "/pdb/admin/v#{@admin_api_version}/archive"
      self.class.post(path, body: { archive: File.open(filename) })
    end

    def status
      status_endpoint = '/status/v1/services'
      status_map = {}

      @servers.each do |server|
        self.class.base_uri(server)
        ret = self.class.get(status_endpoint)

        status_map[server] = if ret.code >= 400
                               { error: "Unable to build JSON object from server: #{server}" }
                             else
                               ret.parsed_response
                             end
      end
      status_map
    end
  end
end
